src/shared/http/Curl.php

```diff
@@ -23,77 +23,97 @@
-        $this->curlHandle = curl_init($url);
+    /** @var array<int, mixed> */
+    private $options = [];
+
+    /** @var int */
+    private $httpCode = 0;
+
+    /** @var string */
+    private $errorMessage = '';
+
+    /** @var int */
+    private $errorCode = 0;
+
+    public function init(string $url): void {
+        $this->url          = $url;
+        $this->options      = [];
+        $this->httpCode     = 0;
+        $this->errorMessage = '';
+        $this->errorCode    = 0;
     }
 
     public function setResolve(string $resolveString): void {
-        curl_setopt($this->curlHandle, CURLOPT_RESOLVE, [$resolveString]);
+        $this->options[CURLOPT_RESOLVE] = [$resolveString];
     }
 
     public function addHttpHeaders(array $headers): void {
-        curl_setopt($this->curlHandle, CURLOPT_HTTPHEADER, $headers);
+        $this->options[CURLOPT_HTTPHEADER] = $headers;
     }
 
     public function disableProgressMeter(): void {
-        curl_setopt($this->curlHandle, CURLOPT_NOPROGRESS, true);
+        $this->options[CURLOPT_NOPROGRESS] = true;
     }
 
     public function doNotReturnBody(): void {
-        curl_setopt($this->curlHandle, CURLOPT_NOBODY, true);
+        $this->options[CURLOPT_NOBODY] = true;
     }
 
     public function enableProgressMeter(callable $progressFunction): void {
-        curl_setopt($this->curlHandle, CURLOPT_NOPROGRESS, false);
-        curl_setopt($this->curlHandle, CURLOPT_PROGRESSFUNCTION, $progressFunction);
-    }
-
-    public function getHttpCode(): int {
-        return (int)curl_getinfo($this->curlHandle, CURLINFO_HTTP_CODE);
+        $this->options[CURLOPT_NOPROGRESS]       = false;
+        $this->options[CURLOPT_PROGRESSFUNCTION] = $progressFunction;
     }
 
     public function setCertificateFile(string $filename): void {
-        curl_setopt($this->curlHandle, CURLOPT_CAINFO, $filename);
+        $this->options[CURLOPT_CAINFO] = $filename;
     }
 
     public function setHeaderFunction(callable $headerFunction): void {
-        curl_setopt($this->curlHandle, CURLOPT_HEADERFUNCTION, $headerFunction);
+        $this->options[CURLOPT_HEADERFUNCTION] = $headerFunction;
     }
 
     public function setOptArray(array $options): void {
-        curl_setopt_array($this->curlHandle, $options);
+        $this->options = $options + $this->options;
     }
 
     /**
      * @throws CurlException
-     * @psalm-suppress InvalidReturnType
      */
     public function exec(): string {
-        $result = curl_exec($this->curlHandle);
+        $ch = curl_init($this->url);
+        assert($ch !== false);
+        curl_setopt_array($ch, $this->options);
+        $result         = curl_exec($ch);
+        $this->httpCode = (int)curl_getinfo($ch, CURLINFO_HTTP_CODE);
 
         if ($result === false) {
+            $this->errorMessage = curl_error($ch);
+            $this->errorCode    = curl_errno($ch);
+
             throw new CurlException('Request failed');
```

Comment:
Maybe, when we already touch this, we should move the error Message and Code into the Exception?

Reply:
That would make the properties and the associated getters useless ,)

Reply:
I like the idea. It's done.